Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg2.cpp

```diff
@@ -0,0 +1,646 @@
+    dataType *a_h = sycl::malloc_host<dataType>(nnzUB, q);
+    dataType *x_h = sycl::malloc_host<dataType>(n, q);
+    dataType *b_h = sycl::malloc_host<dataType>(n, q);
+
+    if (!ia_h || !ja_h || !a_h || !x_h || !b_h ) {
+        throw std::runtime_error("Failed to allocate host side USM memory");
+    }
+
+    // 
+    // Generate a 27 point stencil for 3D laplacian using size elements in each dimension
+    //
+    generate_sparse_matrix<dataType, intType>(size, ia_h, ja_h, a_h);
+
+    // Init right hand side and vector x
+    for (int i = 0; i < n; i++) {
+        b_h[i] = set_fp_value(dataType(1.0), dataType(0.0)); // rhs b = 1
+        x_h[i] = set_fp_value(dataType(0.0), dataType(0.0)); // initial guess x0 = 0
+    }
+
+    //
+    // Execute Preconditioned Conjugate Gradient Algorithm
+    //
+    // solve  A x = b  starting with initial guess x = x0
+    //
+
+    std::cout << "\n\t\tsparse PCG parameters:\n";
+
+    std::cout << "\t\t\tA size: (" << n << ", " << n << ")" << std::endl;
+    std::cout << "\t\t\tPreconditioner = Symmetric Gauss-Seidel" << std::endl;
+    std::cout << "\t\t\tmax iterations = " << maxIter << std::endl;
+    std::cout << "\t\t\trelative tolerance limit = " << relTol << std::endl;
+    std::cout << "\t\t\tabsolute tolerance limit = " << absTol << std::endl;
+
+
+    const intType nnz = ia_h[n]; // assumes zero indexing
+
+    // create arrays for help
+    intType *ia_d    = sycl::malloc_device<intType>(n+1, q);
+    intType *ja_d    = sycl::malloc_device<intType>(nnz, q);
+    dataType *a_d    = sycl::malloc_device<dataType>(nnz, q);
+    dataType *x_d    = sycl::malloc_device<dataType>(n, q);
+    dataType *b_d    = sycl::malloc_device<dataType>(n, q);
+    dataType *r_d    = sycl::malloc_device<dataType>(n, q); // r = b - A*x
+    dataType *z_d    = sycl::malloc_device<dataType>(n, q); // z = M^{-1} * r
+    dataType *p_d    = sycl::malloc_device<dataType>(n, q); // p = ?? z
+    dataType *t_d    = sycl::malloc_device<dataType>(n, q); // helper array
+    dataType *d_d    = sycl::malloc_device<dataType>(n, q); // diag(A)
+    dataType *invd_d = sycl::malloc_device<dataType>(n, q); // inv(diag(A))
+
+    const intType width = 8; // width * sizeof(dataType) >= cacheline size (64 Bytes)
+    dataType *temp_d = sycl::malloc_device<dataType>(4*width, q);
+    dataType *temp_h = sycl::malloc_host<dataType>(4*width, q);
+
+    if ( !ia_d || !ja_d || !a_d || !x_d || !b_d || !z_d || !p_d || !t_d || !d_d || !invd_d || !temp_d || !temp_h) {
+        throw std::runtime_error("Failed to allocate device side USM memory");
+    }
+
+    // device side aliases scattered by width elements each
+    dataType *normr_h  = temp_h;
+    dataType *normr_d  = temp_d;
+    //dataType *rtz_h    = temp_h+1*width;
+    dataType *rtz_d    = temp_d+1*width;
+    //dataType *oldrtz_h = temp_h+2*width;
+    dataType *oldrtz_d = temp_d+2*width;
+    //dataType *pAp_h    = temp_h+3*width;
+    dataType *pAp_d    = temp_d+3*width;
```

Comment:
Do you want to adjust the alloc size on the host and delete the commented lines?
Or is that to show how we skip some syncing points when using device size scalars?